### ISSUE TITLE
Can you provide the download link for the preprocessed dataset?

### ISSUE THREAD

**Issue by IceIce1ce** [Szym29/SUICA#1]

**Comment by Szym29** [comment]
Hi, 

You can find one preprocessed data [here](https://drive.google.com/file/d/1Haglb_e4VwI6BQ1w7E1Z5JSheRNeFo71/view?usp=sharing). For remaning ones, please check the [MOSTA database](https://db.cngb.org/stomics/mosta/download/).

**Comment by IceIce1ce** [comment]
Thank author.

**Comment by StackChan** [comment]
> Hi,
> 
> You can find one preprocessed data [here](https://drive.google.com/file/d/1Haglb_e4VwI6BQ1w7E1Z5JSheRNeFo71/view?usp=sharing). For remaning ones, please check the [MOSTA database](https://db.cngb.org/stomics/mosta/download/).

Also I want to provide another dataset used in the paper to help you reproduce the paper:
10x Visium Human Brain (spatially sparse, 4,910 cells, 36,592 genes, with SIREN)
link: https://www.10xgenomics.com/cn/datasets/adult-human-brain-1-cerebral-cortex-unknown-orientation-stains-anti-gfap-anti-nfh-1-standard-1-1-0

> Adult human cerebral cortex space transcriptome dataset (V1_Human_Brain_Section_1—Adult Human Brain 1, Cerebral Cortex, Unknown orientation) was downloaded from the 10× Genomics. 


Also I appreciate the owner for your fatanstic works. And BTW, may you provide code and seed to generated train/validation dataset so that I can reproduce these 3 tasks? Or else I completely can't reproduce the results.



See #2 also.

**Comment by Szym29** [comment]
Hi @StackChan ,

Please check the config files under /configs/ST. We used 8848 as the random seed for all tasks.

**Comment by StackChan** [comment]
I write reproduce.py as follows to reproduce the results, and follows the instruction in the paper.

Instruction in the paper:
`We apply SUICA to perform degradation-agnostic reconstruction upon ST data under various common degradations, including spatial sparsity, gene drop-out, and noise, to which we refer as spatial imputation, gene imputation and denoising respectively for clarity. Accordingly, we follow different evaluation protocols: for spatial imputation, we randomly sample 80% of the spots for training, leaving the rest 20% for evaluation; for gene imputation, we randomly mute 70% of the elements in the data matrices; for denoising, a standard Gaussian noise is injected to the raw data. Note that, such reconstruction is completely based on the great approximation power and internal smoothness of INRs without any beforehand knowledge of the degradation type in a reference-free manner.`


reproduce.py:

```python
import os
import numpy as np
import torch
import anndata as ad
from omegaconf import OmegaConf
from lightning import pytorch as pl
import importlib

# 导入现有模块
from utils import metrics
from systems import train_embedder, train_inr
from datasets import ST2D as OriginalST2D


def set_seed(seed):
    """强制设置所有随机种子，确保实验可复现"""
    np.random.seed(seed)
    torch.manual_seed(seed)
    pl.seed_everything(seed, workers=True)  # 覆盖lightning的seed


# --------------------------
# 预训练GAE（为gae+inr流程准备）
# --------------------------
def pretrain_gae(case, seed=8848):
    print("=== 开始预训练GAE ===")
    set_seed(seed)
    # 加载GAE配置
    gae_config = OmegaConf.load("configs/ST/embedder_gae.yaml")
    gae_config.case = case
    gae_config.dataset.data_file = f"data/preprocessed_data/{case}.h5ad"
    gae_config.pipeline.optimization.logs = f"logs/GAE/{case}"
    # 训练GAE（生成embedded-all.h5ad）
    train_embedder(gae_config)
    return gae_config


# --------------------------
# 任务1: 空间插补 (gae+inr)
# --------------------------
def run_spatial_interpolation(case, seed=8848, model_type="SIREN"):
    print("\n=== 开始空间插补任务 (gae+inr) ===")
    set_seed(seed)  # 单独设置seed
    
    # 为空间插补任务单独训练GAE（val_proportion=0.2）
    print("=== 为空间插补任务预训练GAE ===")
    gae_config = OmegaConf.load("configs/ST/embedder_gae.yaml")
    gae_config.case = case
    gae_config.dataset.data_file = f"data/preprocessed_data/{case}.h5ad"
    gae_config.dataset.val_proportion = 0.2  # 80%训练，20%验证
    gae_config.pipeline.optimization.logs = f"logs/GAE/{case}_spatial"
    train_embedder(gae_config)
    
    # 加载INR配置（基于GAE的嵌入）
    config = OmegaConf.load("configs/ST/inr_embd.yaml")
    config.case = case
    # 数据路径依赖GAE的输出
    config.dataset.data_file = f"logs/GAE/{case}_spatial/lightning_logs/version_0/embedded-all.h5ad"
    config.pipeline.optimization.logs = f"logs/GAE+{model_type}/{case}_spatial"
    config.pipeline.reconstructed_data = "spatial_reconstructed-val.h5ad"
    # 设置INR模型类型（SIREN或FFN）
    config.pipeline.inr.model = model_type
    # 更新decoder checkpoint路径
    config.pipeline.inr.decoder.ckpt = f"logs/GAE/{case}_spatial/lightning_logs/version_0/checkpoints/last.ckpt"

    # 运行INR训练（前1k epoch训练INR，后1k epoch训练decoder）
    train_inr(config)

    # 计算指标
    result_path = os.path.join(config.pipeline.optimization.logs, 
                             "lightning_logs/version_0", 
                             config.pipeline.reconstructed_data)
    adata = ad.read_h5ad(result_path)
    y_true = adata.X.toarray() if hasattr(adata.X, 'toarray') else adata.X
    y_pred = adata.obsm["reconstructed_raw"].toarray() if hasattr(adata.obsm["reconstructed_raw"], "toarray") else adata.obsm["reconstructed_raw"] # gae+inr的输出在reconstructed_raw
    scores = metrics(y_true, y_pred, prefix="spatial")
    print("[空间插补指标]")
    for k, v in scores.items():
        print(f"{k}: {v:.4f}")
    return scores


# --------------------------
# 任务2: 基因插补 (gae+inr)
# --------------------------
class GeneInterpST2D(OriginalST2D):
    def __getitem__(self, idx):
        base_item = super().__getitem__(idx)
        # 对原始表达数据进行70%静音处理
        raw = base_item["raw_representations"].numpy()
        mask = np.random.choice([True, False], size=raw.shape, p=[0.7, 0.3])
        raw_masked = raw.copy()
        raw_masked[mask] = 0  # 静音处理
        base_item["raw_representations"] = torch.Tensor(raw_masked)
        base_item["raw_gt"] = torch.Tensor(raw)  # 保存真实值
        return base_item

def run_gene_interpolation(case, seed=8848, model_type="SIREN"):
    print("\n=== 开始基因插补任务 (gae+inr) ===")
    set_seed(seed)  # 单独设置seed
    
    # 为基因插补任务单独训练GAE（全量训练，静音处理在后续进行）
    print("=== 为基因插补任务预训练GAE ===")
    gae_config = OmegaConf.load("configs/ST/embedder_gae.yaml")
    gae_config.case = case
    gae_config.dataset.data_file = f"data/preprocessed_data/{case}.h5ad"
    gae_config.dataset.val_proportion = 0.0  # 全量训练
    gae_config.pipeline.optimization.logs = f"logs/GAE/{case}_gene"
    train_embedder(gae_config)
    
    # 临时替换数据集类（注入静音）
    import datasets
    datasets.ST2D = GeneInterpST2D

    # 加载INR配置
    config = OmegaConf.load("configs/ST/inr_embd.yaml")
    config.case = case
    config.dataset.data_file = f"logs/GAE/{case}_gene/lightning_logs/version_0/embedded-all.h5ad"
    config.dataset.val_proportion = 0.0  # 全量训练
    config.pipeline.optimization.logs = f"logs/GAE+{model_type}/{case}_gene"
    config.pipeline.reconstructed_data = "gene_reconstructed-all.h5ad"
    config.pipeline.inr.model = model_type
    # 更新decoder checkpoint路径
    config.pipeline.inr.decoder.ckpt = f"logs/GAE/{case}_gene/lightning_logs/version_0/checkpoints/last.ckpt"

    # 运行INR训练
    train_inr(config)

    # 计算指标（重点关注mask区域）
    result_path = os.path.join(config.pipeline.optimization.logs, 
                             "lightning_logs/version_0", 
                             config.pipeline.reconstructed_data)
    adata = ad.read_h5ad(result_path)
    y_true = adata.X.toarray() if hasattr(adata.X, 'toarray') else adata.X
    y_pred = adata.obsm["reconstructed_raw"].toarray() if hasattr(adata.obsm["reconstructed_raw"], "toarray") else adata.obsm["reconstructed_raw"] # gae+inr的输出在reconstructed_raw

    
    # 计算masked指标（需复用原始mask，这里简化为直接计算非零区域）
    mask = (y_true == 0)  # 假设静音区域为0
    masked_gt = y_true[mask]
    masked_pred = y_pred[mask]
    masked_scores = metrics(masked_gt, masked_pred, prefix="masked")

    scores = metrics(y_true, y_pred, prefix="gene")
    print("[基因插补指标]")
    for k, v in scores.items():
        print(f"{k}: {v:.4f}")
    print("[masked区域指标]")
    for k, v in masked_scores.items():
        print(f"{k}: {v:.4f}")

    # 恢复原始数据集类
    datasets.ST2D = OriginalST2D
    return scores


# --------------------------
# 任务3: 去噪 (gae+inr)
# --------------------------
class DenoiseST2D(OriginalST2D):
    def __getitem__(self, idx):
        base_item = super().__getitem__(idx)
        # 注入高斯噪声
        raw = base_item["raw_representations"].numpy()
        noise = np.random.normal(loc=0, scale=1.0, size=raw.shape)
        raw_noisy = raw + noise
        base_item["raw_representations"] = torch.Tensor(raw_noisy)
        base_item["raw_gt"] = torch.Tensor(raw)  # 保存真实值
        return base_item

def run_denoising(case, seed=8848, model_type="SIREN"):
    print("\n=== 开始去噪任务 (gae+inr) ===")
    set_seed(seed)  # 单独设置seed
    
    # 为去噪任务单独训练GAE（val_proportion=0.2）
    print("=== 为去噪任务预训练GAE ===")
    gae_config = OmegaConf.load("configs/ST/embedder_gae.yaml")
    gae_config.case = case
    gae_config.dataset.data_file = f"data/preprocessed_data/{case}.h5ad"
    gae_config.dataset.val_proportion = 0.2  # 80%训练，20%验证
    gae_config.pipeline.optimization.logs = f"logs/GAE/{case}_denoise"
    train_embedder(gae_config)
    
    # 临时替换数据集类（注入噪声）
    import datasets
    datasets.ST2D = DenoiseST2D

    # 加载INR配置
    config = OmegaConf.load("configs/ST/inr_embd.yaml")
    config.case = case
    config.dataset.data_file = f"logs/GAE/{case}_denoise/lightning_logs/version_0/embedded-all.h5ad"
    config.pipeline.optimization.logs = f"logs/GAE+{model_type}/{case}_denoise"
    config.pipeline.reconstructed_data = "denoise_reconstructed-val.h5ad"
    config.pipeline.inr.model = model_type
    # 更新decoder checkpoint路径
    config.pipeline.inr.decoder.ckpt = f"logs/GAE/{case}_denoise/lightning_logs/version_0/checkpoints/last.ckpt"

    # 运行INR训练
    train_inr(config)

    # 计算指标
    result_path = os.path.join(config.pipeline.optimization.logs, 
                             "lightning_logs/version_0", 
                             config.pipeline.reconstructed_data)
    adata = ad.read_h5ad(result_path)
    y_true = adata.X.toarray() if hasattr(adata.X, 'toarray') else adata.X
    y_pred = adata.obsm["reconstructed_raw"].toarray() if hasattr(adata.obsm["reconstructed_raw"], "toarray") else adata.obsm["reconstructed_raw"] # gae+inr的输出在reconstructed_raw

    scores = metrics(y_true, y_pred, prefix="denoise")
    print("[去噪指标]")
    for k, v in scores.items():
        print(f"{k}: {v:.4f}")

    # 恢复原始数据集类
    datasets.ST2D = OriginalST2D
    return scores


# --------------------------
# 主函数: 依次运行所有流程
# --------------------------
if __name__ == "__main__":
    case = "PCA-E13.5_E1S1.MOSTA"
    seed = 8848  # 所有任务共用此seed
    model_type = "SIREN"  # 可选"FFN"或"SIREN"

    # 依次运行三个任务（每个任务都会独立训练GAE）
    spatial_scores = run_spatial_interpolation(case, seed=seed, model_type=model_type)
    gene_scores = run_gene_interpolation(case, seed=seed, model_type=model_type)
    denoise_scores = run_denoising(case, seed=seed, model_type=model_type)

    print("\n=== 所有任务完成 ===")
    print("各任务MAE对比:", {
        "空间插补": spatial_scores["spatial/mean_absolute_error"],
        "基因插补": gene_scores["gene/mean_absolute_error"],
        "去噪": denoise_scores["denoise/mean_absolute_error"]
    })
```

It requireds 8 GB to run on RTX 4090 24G.

**Comment by StackChan** [comment]
I sincerely thank you once again, @Szym29 , and want to express my deep appreciation for your work.


The **result of reproduce.py on 'spatial imputation' task** when case = "PCA-E13.5_E1S1.MOSTA" is:

```
{
    'val_recon/mean_absolute_error': 0.005714962,
    'val_recon/mean_absolute_error_mask': 0.03397027,
    'val_recon/mean_squared_error': 0.00055881025,
    'val_recon/mean_squared_error_mask': 0.0034005267,
    'val_recon/root_mean_squared_error': 0.012437743,
    'val_recon/cosine_similarity': 0.85041374,
    'val_recon/cosine_similarity_mask': 0.8417895268170554,
    'val_recon/sam': 31.260708,
    'val_recon/iou': 0.8772726715923174,
    'val_recon/pearsonr': 0.8477859665248649,
    'val_recon/spearmanr': 0.5018944622922931,
    'val_recon/pearsonr_mask': 0.8462688828481666,
    'val_recon/spearmanr_mask': 0.5267460854389737
}

```

Although the current results are **based only on the E13.5_E1S1.MOSTA sample** rather than the full Stereo-seq MOSTA dataset, I would greatly appreciate it if the author could kindly **confirm whether my implementation is accurate**.

PLUS：I list results on the **full Stereo-seq MOSTA dataset** in the paper for your convenience:
<img width="553" height="355" alt="Image" src="https://github.com/user-attachments/assets/59b72bbc-94d7-4228-9259-aab6e649beb6" />

**Comment by IceIce1ce** [comment]
Hi @StackChan, I am not the author of this paper. Therefore, I afraid that I cannot help you with this problem. You can contact the author for more detail.

**Comment by Szym29** [comment]
@IceIce1ce 

1. Your reproduction of SUICA results on E13.5 MOSTA slice is valid. 
2. For your code to run benchmarking, spatial imputation and gene imputation should be correct. But to add gaussian noise, be careful that some values may be negative after adding noise which break the assupmtion of ST data. Make sure setting them as 0. Also you need to make sure after adding the noise, the data distribution is still zero-inflated.
